Normalise gateway.conf controls on startup

matthewoliver opened the pull request:
here is a chance that an ses5.5 - 6 upgrade incorrectly stored int
values in the config as strings, and this can lead dashboard issues if
you want to make chances.

This patch will take a look at the config at startup, once it's loaded
from the rados pool, and if there is any ints stored as strings will
write correct them and write it back to the config on the pool.

Signed-off-by: Matthew Oliver <moliver@suse.com>

Commit:
Normalise gateway.conf controls on startup

There is a chance that an ses5.5 - 6 upgrade incorrectly stored int
values in the config as strings, and this can lead dashboard issues if
you want to make chances.

This patch will take a look at the config at startup, once it's loaded
from the rados pool, and if there is any ints stored as strings will
write correct them and write it back to the config on the pool.

While we're fixing ints in the config let's also go and
fix bsc#1163617

Fixes: bsc#1163617
Fixes: bsc#1164571
Signed-off-by: Matthew Oliver <moliver@suse.com>

ricardoasmarques commented on ceph_iscsi_config/common.py:
```diff
@@ -641,6 +641,29 @@ def _commit_rbd(self, post_action):
 
+    def normalise_controls(self):
+        changed = False
+        for section in ('targets', 'disks', 'gateways'):
```
Do we really need `gateways` here? I think we only store `controls` in `/target/<iqn>/controls` and `/disks/<disk_id>/controls`.

matthewoliver replied:
Probably not. Originally I just wanted to loop through all main settings, but then realised everything we needed to change were in controls. But wasn't sure if gateways got them so just left it in.

I'll double check the code and remove `gateways` if not required.

matthewoliver replied:
Looking at the gwcli code, we don't seem to ever set controls in the gateways, so will remove it.

ddiss reviewed (approved):
Looks reasonable. A couple of minor nits in addition to Ricardo's comment:
- please use "bsc#X" bug format so that it's clear that 1163617 is a SUSE ticket (also: where->we're)
- set_item() already sets self.changed, so would it make sense to just check that instead of the extra var?

Will this new normalisation logic only be performed before any (kernel) LIO configuration is done, or could it happen after a target has already been configured? If backstore attributes like "unmap_zeroes_data" are modified via ConfigFS after being exposed (via iSCSI), then the ConfigFS I/O may fail.

matthewoliver commented:
> Looks reasonable. A couple of minor nits in addition to Ricardo's comment:
> 
> * please use "bsc#X" bug format so that it's clear that 1163617 is a SUSE ticket (also: where->we're)
> * set_item() already sets self.changed, so would it make sense to just check that instead of the extra var?

Yeah, good point. I saw that, and for some reason I decided to not to use it.. but for the life of me can't figure out why now, maybe it was before I fully understood the commit logic, no idea.. a blame something :p 
We're using set_item, so yeah, let's use the self.changed :)

> 
> Will this new normalisation logic only be performed before any (kernel) LIO configuration is done, or could it happen after a target has already been configured? If backstore attributes like "unmap_zeroes_data" are modified via ConfigFS after being exposed (via iSCSI), then the ConfigFS I/O may fail.

The normalisation currently happens as the api service starts, so could very well happen pre or post LIO configuration (ie it may get restarted). Maybe this is ok, if it's on a boot?
But maybe this is something to think about. Maybe I need to next delve into LIO :)

ddiss commented:
>     Will this new normalisation logic only be performed before any (kernel) LIO configuration is done, or could it happen after a target has already been configured? If backstore attributes like "unmap_zeroes_data" are modified via ConfigFS after being exposed (via iSCSI), then the ConfigFS I/O may fail.
> 
> The normalisation currently happens as the api service starts, so could very well happen pre or post LIO configuration (ie it may get restarted). Maybe this is ok, if it's on a boot?

If it's on boot then it should be fine, as ConfigFS represents in-memory configuration state.

ricardoasmarques commented:
> > ```
> > Will this new normalisation logic only be performed before any (kernel) LIO configuration is done, or could it happen after a target has already been configured? If backstore attributes like "unmap_zeroes_data" are modified via ConfigFS after being exposed (via iSCSI), then the ConfigFS I/O may fail.
> > ```
> > 
> > 
> > The normalisation currently happens as the api service starts, so could very well happen pre or post LIO configuration (ie it may get restarted). Maybe this is ok, if it's on a boot?
> 
> If it's on boot then it should be fine, as ConfigFS represents in-memory configuration state.

@matthewoliver @ddiss What about moving the normalization code to the same place where ceph-iSCSI upgrades the configuration version when needed [1]? Maybe this will guarantee that the normalization always happens before applying LIO config?

[1] https://github.com/SUSE/ceph-iscsi/blob/ses6/ceph_iscsi_config/common.py#L189

matthewoliver commented:
> @matthewoliver @ddiss What about moving the normalization code to the same place where ceph-iSCSI upgrades the configuration version when needed [1]? Maybe this will guarantee that the normalization always happens before applying LIO config?
> 
> [1] https://github.com/SUSE/ceph-iscsi/blob/ses6/ceph_iscsi_config/common.py#L189

Oh that's a good idea!! I'll move it and test it in my hacked customer env and make sure it says on the tin :)

matthewoliver commented:
It seems _upgrade_config is called when ever config.refresh is called (rather often) and in the config __init__ which happens very early in the script.

Looks like a bunch of LIO and LUN setups, i think, are happening at CephiSCSIGateway's define method so we need to make sure it happens BEFORE:  https://github.com/SUSE/ceph-iscsi/blob/ses6/rbd-target-api.py#L2854

So we can either connect it with _upgrade_config, and it should happen during the Config constructor, but will also be run during other events calling refresh:
```
  Method
      refresh(self)
  Found usages  (31 usages found)
      Unclassified usage  (26 usages found)
          ceph-iscsi  (26 usages found)
                (25 usages found)
                  rbd-target-api.py  (25 usages found)
                      ConfigWatcher  (2 usages found)
                          run  (2 usages found)
                              2807 config.refresh()
                              2816 config.refresh()
                      _client  (1 usage found)
                          2243 config.refresh()
                      _disk  (3 usages found)
                          1153 config.refresh()
                          1215 config.refresh()
                          1298 config.refresh()
                      _gateway  (1 usage found)
                          702 config.refresh()
                      _hostgroup  (2 usages found)
                          2557 config.refresh()
                          2567 config.refresh()
                      _target  (1 usage found)
                          491 config.refresh()
                      _target_disk  (2 usages found)
                          871 config.refresh()
                          944 config.refresh()
                      _targetauth  (1 usage found)
                          1687 config.refresh()
                      _update_client  (1 usage found)
                          1867 config.refresh()
                      disk  (1 usage found)
                          1009 config.refresh()
                      gateway  (2 usages found)
                          603 config.refresh()
                          684 config.refresh()
                      get_config  (1 usage found)
                          508 config.refresh()
                      local_target_reconfigure  (1 usage found)
                          372 config.refresh()
                      lun_reconfigure  (2 usages found)
                          1306 config.refresh()
                          1368 config.refresh()
                      signal_reload  (1 usage found)
                          2910 config.refresh()
                      target  (3 usages found)
                          312 config.refresh()
                          351 config.refresh()
                          356 config.refresh()
              ceph_iscsi_config  (1 usage found)
                  lun.py  (1 usage found)
                      LUN  (1 usage found)
                          allocate  (1 usage found)
                              744 self.config.refresh()
```

Which "should" be fine as it should only update config as required. Or actually leaving it here it is, also makes sure it's only ever run before the define. I guess the problem is, if the service is ever restarted then it'll be run and potentially correct something the is in use by LIO.
Though having said that, if LIO is setup, hopefully that's only because the service has been run in the past, so the changes have already been made.

Except of course when a customer does an update to a live system. When the service is restarted after the update, and they have hit this bug. Then suddenly there could be issues. So an update, may require and reboot, to be sure the LIO isn't loaded (?)

ricardoasmarques commented:
> Or actually leaving it here it is, also makes sure it's only ever run before the define.

Indeed. You normalize the config before applying config to LIO (CephiSCSIGateway's define) so leaving it where it is looks good to me.

> I guess the problem is, if the service is ever restarted then it'll be run and potentially correct something the is in use by LIO.

Note that `ceph-iscsi` will clear the LIO config when `rbd-target-api` service stops, so this scenario will never happen.

> Except of course when a customer does an update to a live system. When the service is restarted after the update, and they have hit this bug. Then suddenly there could be issues. So an update, may require and reboot, to be sure the LIO isn't loaded (?)

Not sure if I understood what you mean here, but when you install a new `ceph-iscsi` package, `rbd-target-api` service is automatically restarted (which means config will be normalized and LIO will be 
 reconfigured).

ricardoasmarques reviewed (approved):
lgtm
